### PR TITLE
rfc5280

### PR DESCRIPTION
**Pull request by felix**
Resurrecting and completing an old pull request by chjj.

### CODE REVIEW

**Comment by indutny** [comment]
Nice! Does it work? ;)

**Comment by felix** [comment]
It seems to ;-)
I have not looked at all the schemas and have only added a test for a string type. May have time to look at it in the next couple of days.

**Comment by indutny** [comment]
Ok, it would be great to test more of these for sure :) Thanks!

**Comment by indutny** [comment]
@felix ping ;)

**Comment by felix** [comment]
@indutny I have not forgotten about this. Project usage will force me to write tests for it soon.

**Comment by felix** [comment]
@indutny I have fleshed out the schemas and RFC 3280's tests work with 5280. The extension parsing tests could be expanded more when time permits.

**Comment by felix** [comment]
@indutny anything more to get this merged?

File: lib/asn1/base/node.js [hunk 1]
```diff
@@ -599,11 +622,33 @@ Node.prototype._encodeChoice = function encodeChoice(data, reporter) {
 Node.prototype._encodePrimitive = function encodePrimitive(tag, data) {
   var state = this._baseState;
 
-  if (tag === 'octstr' || tag === 'bitstr' || tag === 'ia5str')
+  if (tag === 'bitstr')
```
Comment (indutny) [inline lib/asn1/base/node.js]: I think it came to the point where using `/str$/` regexp may be a better than many ifs.

File: lib/asn1/base/node.js [hunk 2]
```diff
@@ -391,11 +392,33 @@ Node.prototype._decodeGeneric = function decodeGeneric(tag, input) {
     return null;
   if (tag === 'seqof' || tag === 'setof')
     return this._decodeList(input, tag, state.args[0]);
-  else if (tag === 'octstr' || tag === 'bitstr')
+  else if (tag === 'bitstr')
```
Comment (indutny) [inline lib/asn1/base/node.js]: Same as below.

File: lib/asn1/decoders/der.js [hunk 3]
```diff
@@ -123,42 +123,55 @@ DERNode.prototype._decodeList = function decodeList(buffer, tag, decoder) {
...
+      return buffer.error('Decoding of string type: bmpstr length mismatch');
+
+    var str = '';
+    for (var i = 0; i < raw.length / 2; i++) {
+      str += String.fromCharCode(raw.readUInt16BE(i * 2));
+    }
+    return str;
+  } else if (tag === 'charstr') {
+    return buffer.raw().toString();
+  } else if (tag === 'genstr') {
```
Comment (indutny) [inline lib/asn1/decoders/der.js]: Let's pus them under a single if condition.

File: rfc/5280/index.js [hunk 4]
```diff
@@ -0,0 +1,857 @@
...
+    this.key('signatureAlgorithm').use(AlgorithmIdentifier),
+    this.key('signature').bitstr()
+  );
+});
+rfc5280.CerficateList = CertificateList;
+
+// AlgorithmIdentifier  ::=  SEQUENCE  {
+//      algorithm               OBJECT IDENTIFIER,
+//      parameters              ANY DEFINED BY algorithm OPTIONAL  }
+var AlgorithmIdentifier = asn1.define('AlgorithmIdentifier', function() {
```
Comment (indutny) [inline rfc/5280/index.js]: Could it possibly reuse some declarations from RFC 3280? As far as I get it, it should be possible.

Comment (indutny) [inline rfc/5280/index.js]: Or maybe we should remove 3280 from this repo at all?

File: rfc/5280/index.js [hunk 5]
```diff
@@ -0,0 +1,857 @@
...
+        certificatePolicies:          CertificatePolicies,
+        policyMappings:               PolicyMappings,
+        authorityKeyIdentifier:       AuthorityKeyIdentifier,
+        policyConstraints:            PolicyConstraints,
+        extendedKeyUsage:             ExtendedKeyUsage,
+        freshestCRL:                  FreshestCRL,
+        inhibitAnyPolicy:             InhibitAnyPolicy,
+        authorityInformationAccess:   AuthorityInfoAccessSyntax,
+        subjectInformationAccess:     SubjectInformationAccess
+      }[obj.extnID]
```
Comment (indutny) [inline rfc/5280/index.js]: Missing semicolon

File: rfc/5280/index.js [hunk 6]
```diff
@@ -0,0 +1,857 @@
...
+// Extension  ::=  SEQUENCE  {
+//      extnID      OBJECT IDENTIFIER,
+//      critical    BOOLEAN DEFAULT FALSE,
+//      extnValue   OCTET STRING }
+var Extension = asn1.define('Extension', function() {
+  this.seq().obj(
+    this.key('extnID').objid(x509OIDs),
+    this.key('critical').bool().def(false),
+    this.key('extnValue').octstr().contains(function (obj) {
+      var out = {
```
Comment (indutny) [inline rfc/5280/index.js]: The hashmap should be declared outside, no need in evaluating it every time.

File: rfc/5280/index.js [hunk 7]
```diff
@@ -0,0 +1,857 @@
...
+        subjectDirectoryAttributes:   SubjectDirectoryAttributes,
```
Comment (indutny) [inline rfc/5280/index.js]: Please don't use this alignment thing, just a single space after `:`.

File: rfc/5280/index.js [hunk 8]
```diff
@@ -0,0 +1,857 @@
...
+
+// Extension  ::=  SEQUENCE  {
+//      extnID      OBJECT IDENTIFIER,
+//      critical    BOOLEAN DEFAULT FALSE,
+//      extnValue   OCTET STRING }
+var Extension = asn1.define('Extension', function() {
+  this.seq().obj(
+    this.key('extnID').objid(x509OIDs),
+    this.key('critical').bool().def(false),
+    this.key('extnValue').octstr().contains(function (obj) {
```
Comment (indutny) [inline rfc/5280/index.js]: No need in space between `function` and `(`, it is not consistent with the rest of the file.

File: rfc/5280/index.js [hunk 9]
```diff
@@ -0,0 +1,857 @@
...
+rfc5280.CertificateSerialNumber = CertificateSerialNumber;
+
+// ORAddress ::= SEQUENCE {
+//    built-in-standard-attributes BuiltInStandardAttributes,
+//    built-in-domain-defined-attributes    BuiltInDomainDefinedAttributes OPTIONAL,
+//    extension-attributes ExtensionAttributes OPTIONAL }
+var ORAddress = asn1.define('ORAddress', function() {
+  this.seq().obj(
+    this.key('builtInStandardAttributes').use(BuiltInStandardAttributes),
+    this.key('builtInDomainDefinedAttributes').optional().use(BuiltInDomainDefinedAttributes),
```
Comment (indutny) [inline rfc/5280/index.js]: 80 column limit?

File: rfc/5280/index.js [hunk 10]
```diff
@@ -0,0 +1,857 @@
...
+//                     (1..ub-domain-defined-attributes) OF BuiltInDomainDefinedAttribute
+var BuiltInDomainDefinedAttributes = asn1.define('BuiltInDomainDefinedAttributes', function() {
+  this.seqof(BuiltInDomainDefinedAttribute);
+});
+rfc5280.BuiltInDomainDefinedAttributes = BuiltInDomainDefinedAttributes;
+
+// BuiltInDomainDefinedAttribute ::= SEQUENCE {
+//    type PrintableString (SIZE (1..ub-domain-defined-attribute-type-length)),
+//    value PrintableString (SIZE (1..ub-domain-defined-attribute-value-length)) }
+var BuiltInDomainDefinedAttribute = asn1.define('BuiltInDomainDefinedAttribute', function() {
```
Comment (indutny) [inline rfc/5280/index.js]: 80 column?

File: rfc/5280/index.js [hunk 11]
```diff
@@ -0,0 +1,857 @@
...
+rfc5280.DeltaCRLIndicator = DeltaCRLIndicator;
+
+// IssuingDistributionPoint ::= SEQUENCE {
+//         distributionPoint          [0] DistributionPointName OPTIONAL,
+//         onlyContainsUserCerts      [1] BOOLEAN DEFAULT FALSE,
+//         onlyContainsCACerts        [2] BOOLEAN DEFAULT FALSE,
+//         onlySomeReasons            [3] ReasonFlags OPTIONAL,
+//         indirectCRL                [4] BOOLEAN DEFAULT FALSE,
+//         onlyContainsAttributeCerts [5] BOOLEAN DEFAULT FALSE }
+var IssuingDistributionPoint = asn1.define('IssuingDistributionPoint', function() {
```
Comment (indutny) [inline rfc/5280/index.js]: 80 column.

File: rfc/5280/package.json [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    "der"
+  ],
+  "author": "Fedor Indutny",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/indutny/asn1.js/issues"
+  },
+  "homepage": "https://github.com/indutny/asn1.js",
+  "peerDependencies": {
+    "asn1.js": "~0.4.0"
```
Comment (indutny) [inline rfc/5280/package.json]: Huh, recent `asn1.js` is `4.4.0`. Let's make it a normal dependency too: `^4.4.0`

File: rfc/5280/test/chain-test.js [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+var Buffer = require('buffer').Buffer;
+
+describe('asn1.js RFC5280', function() {
+  it('should decode Certificate Extensions', function() {
+    var chain = [];
+    var cert;
+    var extensions = {}
+
+    chain[0] = new Buffer(''
```
Comment (indutny) [inline rfc/5280/test/chain-test.js]: I was thinking about it lately, what do you think about putting this blobs of data into `test/fixtures/` as just files, and reading them with `fs` here?

**Comment by indutny** [comment]
Sorry for delay @felix ! Left some comments, thank you!

File: rfc/5280/index.js (hunk 4 above)
Comment (felix) [inline rfc/5280/index.js]: Perhaps, since 5280 makes 3280 obsolete.

**Comment by felix** [comment]
@indutny cleaned it all up a little except for the blob files, your call on that one.

File: lib/asn1/base/node.js [hunk 14]
```diff
@@ -599,11 +596,9 @@ Node.prototype._encodeChoice = function encodeChoice(data, reporter) {
 Node.prototype._encodePrimitive = function encodePrimitive(tag, data) {
   var state = this._baseState;
 
-  if (tag === 'octstr' || tag === 'bitstr' || tag === 'ia5str')
-    return this._encodeStr(data, tag);
-  else if (tag === 'utf8str' || tag === 'bmpstr')
+  if (tag === 'bitstr')
```
Comment (indutny) [inline lib/asn1/base/node.js]: I guess condition below should handle this too?

File: rfc/5280/test/basic-test.js [hunk 15]
```diff
@@ -0,0 +1,314 @@
+var assert = require('assert');
+var asn1 = require('../../../');
+var rfc5280 = require('..');
+
+var Buffer = require('buffer').Buffer;
+
+describe('asn1.js RFC5280', function() {
+
+  it('should decode Certificate', function() {
+    var data = new Buffer(
```
Comment (indutny) [inline rfc/5280/test/basic-test.js]: Yeah, let's put them into `test/fixtures/` dir.

**Comment by indutny** [comment]
Looks cool, almost ready to land. Let's move the blobs to fixtures dir, move existing 3280 tests to 5280 tests, and remove 3280 module. Should be good to go after that!

**Comment by indutny** [comment]
Thank you!

**Comment by felix** [comment]
Ping @indutny, anything more?

**Comment by indutny** [comment]
@felix there is no github notifications when you push more commits to PR (please make sure to ping me if you want me to take a look). I will take a look at this PR tomorrow!

**Comment by indutny** [comment]
Thank you!

**Comment by indutny** [comment]
Landed and published, thank you!

**Pull request closed without merging**